File: src/main/java/com/lykke/matching/engine/daos/bitcoin/BtTransaction.java
package com.lykke.matching.engine.daos.bitcoin;

import com.google.gson.Gson;
import com.microsoft.azure.storage.table.TableServiceEntity;

import java.util.Date;

public class BtTransaction extends TableServiceEntity {

    public static String TRANS_ID = "TransId";

    String contextData;
    Date created = new Date();
    String requestData;

    public BtTransaction(String rowKey, Date created, String requestData, ClientCashOperationPair clientCashOperationPair) {
        this(rowKey, created, requestData, clientCashOperationPair, null);
    }

    public BtTransaction(String rowKey, Date created, String requestData, Orders orders) {
        this(rowKey, created, requestData, null, orders);
    }

    public BtTransaction(String rowKey, Date created, String requestData, ClientCashOperationPair clientCashOperationPair, Orders orders) {
        super(TRANS_ID, rowKey);
        this.created = created;
        this.requestData = requestData;

        if (clientCashOperationPair != null) {
            setClientCashOperationPair(clientCashOperationPair);
        }
        if (orders != null) {
            setOrders(orders);
        }
    }

    public void setOrders(Orders orders) {
        this.contextData = new Gson().toJson(orders);
    }

    public void setClientCashOperationPair(ClientCashOperationPair clientCashOperationPair) {
        this.contextData = new Gson().toJson(clientCashOperationPair);
    }

    public String getContextData() {
        return contextData;
    }

    public void setContextData(String contextData) {
        this.contextData = contextData;
    }

    public Date getCreated() {
        return created;
    }

    public void setCreated(Date created) {
        this.created = created;
    }

    public String getRequestData() {
        return requestData;
    }

    public void setRequestData(String requestData) {
        this.requestData = requestData;
    }
}

File: src/main/kotlin/com/lykke/matching/engine/daos/BtTransaction.kt
package com.lykke.matching.engine.daos

import com.google.gson.Gson
import com.microsoft.azure.storage.table.TableServiceEntity
import java.util.Date

class BtTransaction : TableServiceEntity {

    var contextData: String? = null
    var created = Date()
    var requestData: String?

    constructor(partitionKey: String?, rowKey: String?, created: Date, requestData: String,
                clientCashOperationPair: ClientCashOperationPair? = null, orders: Orders? = null) : super(partitionKey, rowKey) {
        this.created = created
        this.requestData = requestData
        if (clientCashOperationPair != null) {
            addClientCashOperationPair(clientCashOperationPair)
        }
        if (orders != null) {
            addOrders(orders)
        }
    }

    fun addOrders(orders: Orders) {
        this.contextData = Gson().toJson(orders)
    }

    fun addClientCashOperationPair(clientCashOperationPair: ClientCashOperationPair) {
        this.contextData = Gson().toJson(clientCashOperationPair)
    }
}

class ClientOrderPair(val ClientId: String, val OrderId: String)
class ClientTradePair(val ClientId: String, val TradeId: String)
class Orders(val MarketOrder: ClientOrderPair, val LimitOrder: ClientOrderPair, val Trades: Array<ClientTradePair>)
class ClientCashOperationPair(val ClientId: String, val CashOperationId: String)